src/main/java/edu/umich/its/cpm/MigrationInstanceService.java

```diff
@@ -104,45 +161,113 @@ private void updateMigrationStatusAndEndTime() {
+			// all the items within the migration is finished
+			// update the end time of the parent record
+			Timestamp lastItemMigrationTime = fRepository.getLastItemEndTimeForMigration(mId);
+			mRepository.setMigrationEndTime(lastItemMigrationTime, mId);
+			
+			// update the status of the parent record
+			List<MigrationBoxFile> mFileList = fRepository.getAllItemStatusForMigration(mId);
+			// parse the string into JSON object
+			List<MigrationFileItem> itemStatusList = new ArrayList<MigrationFileItem>();
+			int itemStatusFailureCount = 0;
+			for(MigrationBoxFile mFile : mFileList)
+			{
+				String status = mFile.getStatus();
+				MigrationFileItem item = new MigrationFileItem(
+						mFile.getFile_access_url(), 
+						mFile.getTitle(), 
+						status);
+				itemStatusList.add(item);
 				
-				// update the status of the parent record
-				List<MigrationBoxFile> mFileList = fRepository.getAllItemStatusForMigration(mId);
-				// parse the string into JSON object
-				List<MigrationFileItem> itemStatusList = new ArrayList<MigrationFileItem>();
-				int itemStatusFailureCount = 0;
-				for(MigrationBoxFile mFile : mFileList)
+				if (status.indexOf("Box upload successful for file") == -1)
 				{
-					String status = mFile.getStatus();
-					MigrationFileItem item = new MigrationFileItem(
-							mFile.getFile_access_url(), 
-							mFile.getTitle(), 
-							status);
-					itemStatusList.add(item);
-					
-					if (status.indexOf("Box upload successful for file") == -1)
-					{
-						// if there is error, status message won't have String "Box upload successful for file"
-						itemStatusFailureCount++;
-					}
+					// if there is error, status message won't have String "Box upload successful for file"
+					itemStatusFailureCount++;
 				}
-				
-				// the HashMap object holds itemized status information
-				HashMap<String, Object> statusMap = new HashMap<String, Object>();
-				statusMap.put(Utils.MIGRATION_STATUS, itemStatusFailureCount == 0? Utils.STATUS_SUCCESS:Utils.STATUS_FAILURE);
-				statusMap.put(Utils.MIGRATION_DATA, itemStatusList);
+			}
+			
+			// the HashMap object holds itemized status information
+			HashMap<String, Object> statusMap = new HashMap<String, Object>();
+			statusMap.put(Utils.MIGRATION_STATUS, itemStatusFailureCount == 0? Utils.STATUS_SUCCESS:Utils.STATUS_FAILURE);
+			statusMap.put(Utils.MIGRATION_DATA, itemStatusList);
 
-				// update the status and end_time of migration record
-				migrationTaskService.setMigrationEndTimeAndStatus(mId, mRepository, new JSONObject(statusMap));
+			// update the status of migration record
+			mRepository.setMigrationStatus((new JSONObject(statusMap)).toString(), mId);
+			
+		}
+	}
+	
+	/**
+	 * if all message items within the migration is finished
+	 * update the migration end time with the last end time of items
+	 * update the migration status with the aggregation of item status
+	 * @param mId
+	 */
+	private void updateMessageMigrationTimeAndStatus(String mId) {
```

Comment:
Both time and status methods are quite similar and the common processing should be extracted to a single method.